hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/OzoneClientCache.java

```diff
@@ -75,8 +99,66 @@ public static OzoneClient getOzoneClientInstance(OzoneConfiguration
     return instance.client;
   }
 
+  public static void closeClient() throws IOException {
+    if (instance != null) {
+      instance.client.close();
+      instance = null;
+    }
+  }
+
+  private void setCertificate(String omServiceID,
+                              OzoneConfiguration conf)
+      throws IOException {
+
+    // create local copy of config incase exception occurs
+    // with certificate OmRequest
+    OzoneConfiguration config = new OzoneConfiguration(conf);
+    OzoneClient certClient;
+
+    if (secConfig.isGrpcTlsEnabled()) {
+      // set OmTransport to hadoop rpc to securely,
+      // get certificates with service list request
+      config.set(OZONE_OM_TRANSPORT_CLASS,
+          OZONE_OM_TRANSPORT_CLASS_DEFAULT);
+
+      if (omServiceID == null) {
+        certClient = OzoneClientFactory.getRpcClient(config);
+      } else {
+        // As in HA case, we need to pass om service ID.
+        certClient = OzoneClientFactory.getRpcClient(omServiceID,
+            config);
+      }
+      try {
+        ServiceInfoEx serviceInfoEx = certClient
+            .getObjectStore()
+            .getClientProxy()
+            .getOzoneManagerClient()
+            .getServiceInfo();
+
+        if (OzoneSecurityUtil.isSecurityEnabled(conf)) {
+          String caCertPem = null;
+          List<String> caCertPems = null;
+          caCertPem = serviceInfoEx.getCaCertificate();
+          caCertPems = serviceInfoEx.getCaCertPemList();
+          if (caCertPems == null || caCertPems.isEmpty()) {
+            caCertPems = Collections.singletonList(caCertPem);
```

Comment:
It is possible that caCertPem could also be null.

Reply:
The same conditions exist in the `RpcClient` constructor (RpcClient.java) where `caCertPem` and `caCertPems` are used in a similar check to see if `caCertPems` exists to set the `caCertPems` to `caCertPem` if necessary.  Not sure if we need to address this.  Are we guaranteed to have at least one of them valid and non null (caCertPem)?

Reply:
I don't think it is guaranteed. I commented about it as I am not sure how SslContextBuilder handles a list with null value and if we need to handle that case from our end.

Reply:
Thanks @hanishakoneru for following up.  Yes, I agree - looking at the code for the certificates, it may not be set in neither the `CaCertPemList` nor the `CaCertificate`.  I've added an additional check for null CaCertificate and throws if the not null precondition is not met in the `OzoneClientCache` prior to setCaCerts for the transport.

As a result we handle this with the client creation jax injection error 500 http response (max retries) with an exception trace indicating the precondition failure of not null check on the certificate:  
_java.lang.NullPointerException
at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:880)
,org.apache.hadoop.ozone.s3.OzoneClientCache.setCertificate(OzoneClientCache.java:146)_

Do we need to do anything in addition?

The same conditional check of `caCertPems` for null or empty then setting the `caCertPerms` with the `caCertPer` exists in the `RpcClient` constructor code.  I can followup on this with new jira to patch the `RpcClient` if you think it is needed(?).

Reply:
I don't think the Precondtion check is the right over here as there could be a case where caCertPems is null. 
When caCertPems is null, the request would fail in authentication step anyway, right? If so, we should just log an error here. Throwing a NullPointerException does not look right.

Reply:
Thanks @hanishakoneru for following up on this null check.  Yes, we were discussing the case where `caCertPems` is null and if it would be guaranteed to have at least one of `caCertPems` or `caCertPem` to be non null.  We came up with that we can't make the assumption so to handle that case here ourselves.

In this case with the Precondition check, it is after we determine that `caCertPems` is empty or null (`caCertPem` can't be null here).  In this case we need the `caCertPem` to be non-null or else it causes a NPE in the following `OzoneSecurityUtil.convertToX509 `call.  I put in the precondition check to catch this null violation to intentionally throw an exception so the caller knows that in this case the `caCertPem` must be non null.

What do think if we, 
i.) update the code to check the precondition that `caCertPem` is not null in the case that `caCertPems` is either empty or null with a message such as,
   `Preconditions.checkNotNull(caCertPem, "caCertPem can not be null when caCertPems is empty or null");`
_OR_,
ii.) log the error and throw an exception.

However if we just log the error (`caCertPem` is null) and continue, unfortunately the call that follows to `OzoneSecurityUtil.convertToX509` will throw a NPE.

Reply:
I think logging and throwing an exception would be better than throwing a NullPointerException.

Reply:
Or I am ok with leaving it as before (no extra check, similar to how it is in RpcClient) and we can file a separate Jira to address this at both the places.

Reply:
Thanks @hanishakoneru for your comments on handling the NPE should both the `caCertPems` (list) be empty / null and the `caCertPem` be null. 
I've updated the patch to log the error and throw a `CertificateException` should we have this case.
We can also add this type of exception handling in the RpcClient, or bud it where the NPE is found: `OzoneSecurityUtil.java` `convertToX509` method.  We can file that separate jira for that.  Let me know your thoughts on this.